src/Aeson.purs

```diff
@@ -112,24 +165,156 @@ getField' decoder obj str =
-  decodeAeson (Aeson { index, json })
-  where
-  patchedPayload /\ index = parseJsonExtractingIntegers Tuple payload
+decodeAesonString = parseJsonStringToAeson >=> decodeAeson
+
+-------- DecodeAeson instances --------
+
+instance DecodeAeson Int where
+  decodeAeson aeson@(Aeson { numberIndex }) = do
+    -- Numbers are replaced by their index in the array.
+    ix <- decodeAesonViaJson aeson
+    numberStr <- note MissingValue (numberIndex Array.!! ix)
+    note MissingValue $ Int.fromString numberStr
+
+instance DecodeAeson BigInt where
+  decodeAeson aeson@(Aeson { numberIndex }) = do
+    -- Numbers are replaced by their index in the array.
+    ix <- decodeAesonViaJson aeson
+    numberStr <- note MissingValue (numberIndex Array.!! ix)
+    note MissingValue $ BigInt.fromString numberStr
+
+instance DecodeAeson Boolean where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson String where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson Number where
+  decodeAeson = decodeAesonViaJson
+
+instance DecodeAeson Json where
+  decodeAeson = Right <<< toStringifiedNumbersJson
+
+instance DecodeAeson Aeson where
+  decodeAeson = pure
+
+instance (GDecodeAeson row list, RL.RowToList row list) => DecodeAeson (Record row) where
+  decodeAeson json =
+    case toObject json of
+      Just object -> gDecodeAeson object (Proxy :: Proxy list)
+      Nothing -> Left $ TypeMismatch "Object"
+
+else instance (Traversable t, DecodeAeson a, DecodeJson (t Json)) => DecodeAeson (t a) where
+  decodeAeson (Aeson { numberIndex, patchedJson: AesonPatchedJson pJson }) = do
+    jsons :: t _ <- map AesonPatchedJson <$> decodeJson pJson
+    for jsons (\patchedJson -> decodeAeson (Aeson { patchedJson, numberIndex }))
+
+class GDecodeAeson (row :: Row Type) (list :: RL.RowList Type) | list -> row where
+  gDecodeAeson :: forall proxy. FO.Object Aeson -> proxy list -> Either JsonDecodeError (Record row)
+
+instance GDecodeAeson () RL.Nil where
+  gDecodeAeson _ _ = Right {}
+
+instance
+  ( DecodeAesonField value
+  , GDecodeAeson rowTail tail
+  , IsSymbol field
+  , Row.Cons field value rowTail row
+  , Row.Lacks field rowTail
+  ) =>
+  GDecodeAeson row (RL.Cons field value tail) where
+  gDecodeAeson object _ = do
+    let
+      _field = Proxy :: Proxy field
+      fieldName = reflectSymbol _field
+      fieldValue = FO.lookup fieldName object
+
+    case decodeAesonField fieldValue of
+      Just fieldVal -> do
+        val <- lmap (AtKey fieldName) fieldVal
+        rest <- gDecodeAeson object (Proxy :: Proxy tail)
+        Right $ Record.insert _field val rest
+
+      Nothing ->
+        Left $ AtKey fieldName MissingValue
+
+class DecodeAesonField a where
+  decodeAesonField :: Maybe Aeson -> Maybe (Either JsonDecodeError a)
+
+instance DecodeAeson a => DecodeAesonField (Maybe a) where
+  decodeAesonField Nothing = Just $ Right Nothing
+  decodeAesonField (Just j) = Just $ decodeAeson j
```

Comment:
We could write this as `decodeAesonField = Just <<< maybe (Right Nothing) decodeAeson` although there may be a preference for more explicit class instances to aid readability.